### PR TITLE
Improve parquet partition_file output display

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/4466

# Rationale for this change

See https://github.com/apache/arrow-datafusion/issues/4466

# What changes are included in this PR?
Improve the formatting for `Vec<Vec<PartitionFile>>`

# Are these changes tested?

Yes

# Are there any user-facing changes?
better explain output

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 1]
```diff
@@ -183,17 +183,27 @@ struct FileGroupsDisplay<'a>(&'a [Vec<PartitionedFile>]);
...
-                pp.iter()
-                    .map(|pf| pf.object_meta.location.as_ref())
-                    .collect::<Vec<_>>()
-                    .join(", ")
-            })
-            .collect();
-        write!(f, "[{}]", parts.join(", "))
+        let mut first_group = true;
+
+        for (i, group) in self.0.iter().enumerate() {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/mod.rs:188]: This implementation also avoids allocating strings and Vecs, which I don't think is a particularly important feature, but it is a nice benefit.

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 2]
```diff
@@ -796,4 +808,38 @@ mod tests {
...
+
+    #[test]
+    fn file_groups_display() {
+        let files = vec![
+            vec![partitioned_file("foo"), partitioned_file("bar")],
+            vec![partitioned_file("baz")],
+            vec![],
+        ];
+
+        let expected = "group 0: [foo, bar], group 1: [baz], group 2: []";
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/mod.rs:826]: Prior to the changes in this PR, this was displayed as `"[foo, bar, baz, ]"`


I think this PR is much clearer

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 3]
```diff
@@ -796,4 +808,38 @@ mod tests {
...
     }
+
+    #[test]
+    fn file_groups_display_empty() {
+        let expected = "";
+        assert_eq!(&FileGroupsDisplay(&vec![]).to_string(), expected);
+    }
+
+    #[test]
+    fn file_groups_display() {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/mod.rs:819]: Prior to the changes in this PR, this was displayed as `"[foo, bar, baz, ]"`


I think this PR is much clearer

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 4]
```diff
@@ -183,17 +183,27 @@ struct FileGroupsDisplay<'a>(&'a [Vec<PartitionedFile>]);
 
 impl<'a> Display for FileGroupsDisplay<'a> {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/mod.rs:184]: This implementation also avoids allocating strings and Vecs, which I don't think is a particularly important feature, but it is a nice benefit.

File: datafusion/core/tests/sql/explain_analyze.rs [hunk 5]
```diff
@@ -696,7 +696,7 @@ async fn test_physical_plan_display_indent() {
         "                CoalesceBatchesExec: target_batch_size=4096",
         "                  FilterExec: c12@1 < 10",
         "                    RepartitionExec: partitioning=RoundRobinBatch(9000)",
-        "                      CsvExec: files=[ARROW_TEST_DATA/csv/aggregate_test_100.csv], has_header=true, limit=None, projection=[c1, c12]",
+        "                      CsvExec: files=1:[[ARROW_TEST_DATA/csv/aggregate_test_100.csv]], has_header=true, limit=None, projection=[c1, c12]",
```
Comment (alamb) [inline datafusion/core/tests/sql/explain_analyze.rs:699]: I tried several variations on this formatting to show the file names as well as the number of partitions

Other thoughts?
maybe adding a brace and a key?


```suggestion
        "                      CsvExec: files={groups:1, [[ARROW_TEST_DATA/csv/aggregate_test_100.csv]]}, has_header=true, limit=None, projection=[c1, c12]",
```

**Review by alamb — COMMENTED** [review]
cc @rdettai  who I believe contributed the original formatting code

File: datafusion/core/tests/sql/explain_analyze.rs (hunk 5 above)
Comment (crepererum) [reply datafusion/core/tests/sql/explain_analyze.rs:699]: Since you only spell out `groups` once, I think it could be very helpful to spell this out.

**Review by crepererum — APPROVED** [review]
Like this very much. Would prefer the slightly more verbose version, but that's not a blocker.

**Review by tustvold — APPROVED** [review]
I like the idea of explicitly identifying what the integer represents, the partition count, but equally this is an improvement already so approving 👍

**Comment by alamb** [comment]
I have changed it to be 

```
partitions={1 group: [[x]]}
```

Which I think is better

**Review by xudong963 — APPROVED** [review]
Let's merge! A nice improvement

**Pull request merged**